files/image_config/monit/container_checker

```diff
@@ -0,0 +1,111 @@
+              The 'state' field of container in 'FEATURE' table should not be 'disabled'. Then
+              if the device has Multi-ASIC, this function will get container list by determining the
+              value of field 'has_global_scope', the number of ASICs and the value of field
+              'has_per_asic_scope'. If the device has single ASIC, the container name was put into
+              the list.
+    @return:  A set which contains the expected running containers.
+    """
+    config_db = swsssdk.ConfigDBConnector()
+    config_db.connect()
+    feature_table = config_db.get_table("FEATURE")
+
+    expected_running_containers = set()
+
+    for container_name in feature_table.keys():
+        if feature_table[container_name]["state"] != "disabled":
+            if multi_asic.is_multi_asic():
+                if feature_table[container_name]["has_global_scope"] == "True":
+                    expected_running_containers.add(container_name)
+                if feature_table[container_name]["has_per_asic_scope"] == "True":
+                    num_asics = multi_asic.get_num_asics()
+                    for asic_id in range(num_asics):
+                        expected_running_containers.add(container_name + str(asic_id))
+            else:
+                expected_running_containers.add(container_name)
+
+    return expected_running_containers
+
+
+def get_current_running_containers():
+    """
+    @summary: This function will get the current running container list by analyzing the
+              output of command `docker ps`.
+    @return:  A set which contains the current running contianers.
+    """
+    running_containers = set()
+
+    command = "docker ps"
+    command_stdout = get_command_result(command)
+    for line in command_stdout:
+        if line and "CONTAINER ID" not in line:
+            running_containers.add(line.split()[-1].strip())
+
+    return running_containers
+
+
+def main():
+    """
+    @summary: This function will compare the difference between the current running containers
+              and the containers which were expected to run. If containers which were exepcted
+              to run were not running, then an alerting message will be written into syslog.
+    """
+    expected_running_containers = get_expected_running_containers()
+    current_running_containers = get_current_running_containers()
+
+    not_running_containers = expected_running_containers.difference(current_running_containers)
+
+    if not_running_containers:
+        log_message = ""
+        for container_name in not_running_containers:
+            if not log_message:
+                log_message = container_name
+            else:
+                log_message += ", " + container_name
+
+        if len(not_running_containers) > 1:
+            print("Containers '{}' were not running.".format(log_message))
+        else:
+            print("Container '{}' was not running.".format(log_message))
+
+        sys.exit(4)
```

Comment:
This can be greatly simplified to something like:

```python3
    if not_running_containers:
        print("Containers not running: " + ", ".join(not_running_containers))
        sys.exit(4)
```

Also, suggest logging to syslog rather than simply printing to stdout. You can use the Logger class in the sonic_py_common package. There are multiple examples in the codebase.

Reply:
Updated! The string in the `print(...)` statement will be appended to Monit syslog message. But I like idea to simplify the logic about generating the alerting message.

Reply:
So the string in the `print(...)` statement will get appended to the monit syslog message? Is that the behavior we want? If so, then maybe we shouldn't use the Logger class. It just depends if we expect the message to be logged by Monit or not.

Reply:
Actually in the script `process_checker` we used the `print(...)` statement to write the alerting messages into syslog. 

Log messages were generated by `print(...)` statement:
`Jan  5 02:49:39.432635 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- Containers not running: lldp, pmon`
`Jan  5 02:50:39.917866 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- Containers not running: pmon, lldp`
`Jan  5 02:51:40.413938 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- Containers not running: lldp, pmon`
`Jan  5 02:52:40.948582 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- Containers not running: pmon, lldp`
`Jan  5 02:53:41.420400 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- Containers not running: pmon, lldp`

Log messages were generated by Logger class:
`Jan  5 02:58:43.926697 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- no output`
`Jan  5 02:58:46.635607 str-a7050-acs-3 ERR container_checker.py: Containers not running: pmon, lldp`
`Jan  5 02:59:44.390015 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- no output`
`Jan  5 02:59:47.194435 str-a7050-acs-3 ERR container_checker: Containers not running: lldp, pmon`
`Jan  5 03:00:44.863422 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- no output`
`Jan  5 03:00:47.784087 str-a7050-acs-3 ERR container_checker: Containers not running: lldp, pmon`
`Jan  5 03:01:45.394784 str-a7050-acs-3 ERR monit[19018]: 'container_checker' status failed (4) -- no output`
`Jan  5 03:01:48.015010 str-a7050-acs-3 ERR container_checker: Containers not running: pmon, lldp`

Reply:
Hmm. Then maybe we should revert back to the `print(...)` statements to consolidate all the output into the same line and also prevent the `no output` messages.

Reply:
I agree and will revert back to use the `print(...)` statement.

Reply:
Updated!